File: application/service/blocking/src/main/java/org/occurrent/application/converter/CloudEventConverter.java
package org.occurrent.application.converter;

import io.cloudevents.CloudEvent;

/**
 * A cloud event converter interface that is used by Occurrent application services
 * to convert to and from domain events.
 *
 * @param <T> The type of your domain event
 */
public interface CloudEventConverter<T> {

    /**
     * Convert a domain event into a cloud event
     *
     * @param domainEvent The domain event to convert
     * @return The {@link CloudEvent} instance, converted from the domain event.
     */
    CloudEvent toCloudEvent(T domainEvent);

    /**
     * Convert a cloud event to a domain event
     *
     * @param cloudEvent The cloud event to convert
     * @return The domain event instance, converted from the cloud event.
     */
    T toDomainEvent(CloudEvent cloudEvent);
}

File: example/domain/word-guessing-game/mongodb/spring/blocking/src/main/kotlin/org/occurrent/example/domain/wordguessinggame/mongodb/spring/blocking/application/CloudEventConverter.kt
package org.occurrent.example.domain.wordguessinggame.mongodb.spring.blocking.application

import com.fasterxml.jackson.databind.ObjectMapper
import io.cloudevents.CloudEvent
import org.occurrent.example.domain.wordguessinggame.event.DomainEvent
import java.net.URI

class CloudEventConverter(val objectMapper: ObjectMapper, cloudEventSource: URI) {
    fun toCloudEvent(domainEvent: DomainEvent): CloudEvent = TODO()
    fun toDomainEvent(cloudEvent: CloudEvent): DomainEvent = TODO()
}